File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <bits/stdc++.h>
using namespace std;

int T, n, m, K, x, y;
const int dx[] = {1, -1, 0, 0};
const int dy[] = {0, 0, 1, -1};
int arr[51][51];
int visited[51][51];
int ret;

void dfs(int row, int col){
    visited[row][col] = 1;
    for(int i = 0; i < 4; i++){
        int nx = row + dx[i];
        int ny = col + dy[i];

        if(0 <= nx && nx < n && 0 <= ny && ny < m && !visited[nx][ny] && arr[nx][ny] == 1)
            dfs(nx, ny);

    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(NULL);

    cin >> T;
    for(int t = 0; t < T; t++){
        cin >> m >> n >> K;

        fill(&arr[0][0], &arr[0][0] + 51 * 51, 0);
        fill(&visited[0][0], &visited[0][0] + 51 * 51, 0);
        for(int k = 0; k < K; k++){
            cin >> y >> x;
            arr[x][y] = 1;
        }

        ret = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(!visited[i][j] && arr[i][j] == 1){
                    dfs(i, j);
                    ret++; 
                }
                    
                
            }
        }

        cout << ret << "\n";
    }
    
    return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10000)  # 파이썬 재귀 깊이 제한은 1000

def dfs(row, col):
    visited[row][col] = 1
    for dx, dy in zip([1, -1, 0, 0], [0, 0, 1, -1]):
        nx, ny = row + dx, col + dy

        if nx < 0 or nx >= n or ny < 0 or ny >= m: continue
        if not visited[nx][ny] and grid[nx][ny] == 1:
            dfs(nx, ny)



t = int(input())    
for _ in range(t):
    n, m, k = map(int, input().split())
    grid = [[0] * m for _ in range(n)]
    visited = [[0] * m for _ in range(n)]

    # 배추 심기
    for _ in range(k):
        r, c = map(int, input().split())
        grid[r][c] = 1
    
    # flood fill
    ans = 0 
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and grid[i][j] == 1:  # 해당 칸이 배추가 있고 방문하지 않았으면 dfs실행
                dfs(i, j)
                ans += 1  # dfs가 끝났다? flood fill 한 칸 완료

    print(ans)      
